holopy/inference/interface.py

```diff
@@ -74,36 +74,59 @@ def validate_strategy(strategy, operation):
-        fitting_parameters = base_scatterer.parameters.keys()
+        fitting_parameters = list(base_scatterer.parameters.keys())
     scatterer = parameterize_scatterer(base_scatterer, fitting_parameters)
     alpha_prior = Uniform(0, 1, guess=0.7, name='alpha')
     return AlphaModel(scatterer, noise_sd=1, alpha=alpha_prior)
 
 
 def parameterize_scatterer(base_scatterer, fitting_parameters):
-    parameters = base_scatterer.guess.parameters
+    if isinstance(fitting_parameters, str):
+        fitting_parameters = [fitting_parameters]
+    parameters = base_scatterer.parameters
     variable_parameters = {par_name: make_uniform(parameters, par_name)
-                           for par_name in rename_xyz(fitting_parameters)}
+                           for par_name in fitting_parameters}
     parameters.update(variable_parameters)
-    return type(base_scatterer)(**_interpret_parameters(parameters, True))
-
-
-def rename_xyz(parameters_list):
-    for i, key in enumerate(['x', 'y', 'z']):
-        new_key = 'center.{}'.format(i)
-        if key in parameters_list:
-            loc = parameters_list.index(key)
-            parameters_list[loc] = new_key
-    return parameters_list
+    for i, key in enumerate(COORD_KEYS):
+        if isinstance(base_scatterer, Scatterers):
+            for j in range(len(base_scatterer.scatterers)):
+                replace_center(parameters, "{}:{}".format(j, key))
+        else:
+            replace_center(parameters, key)
+    return base_scatterer.from_parameters(parameters)
+
+
+def replace_center(parameters, key):
+    index = COORD_KEYS.index(key[-1])
+    prefix = key[:-1]
+    try:
+        val = parameters.pop(key)
+    except KeyError:
+        pass
+    else:
+        parameters[prefix + 'center'][index] = val
 
 
 def make_uniform(guesses, key):
-    try:
-        guess_value = guesses[key]
-    except KeyError:
-        msg = 'Parameter {} not found in scatterer parameters.'.format(key)
-        raise ValueError(msg)
-    minval = 0 if key in ['n', 'r'] else -np.inf
-    return Uniform(minval, np.inf, guess_value, key)
+    suffix = key.split(":")[-1]
+    prefix = key[:-len(suffix)]
+    if suffix in COORD_KEYS:
+        guess_value = guesses[prefix + 'center'][COORD_KEYS.index(suffix)]
+    else:
+        try:
+            guess_value = guesses[key]
+        except KeyError:
+            msg = 'Parameter {} not found in scatterer parameters.'.format(key)
+            raise ValueError(msg)
+    minval = 0 if suffix in ['n', 'r'] else -np.inf
+    if isinstance(guess_value, (list, tuple, np.ndarray)):
+        if suffix == 'center':
+            subkeys = COORD_KEYS
+        else:
+            subkeys = [suffix + '.' + str(i) for i in range(len(guess_value))]
+        return [Uniform(minval, np.inf, guess_val, prefix + subkey)
+                for guess_val, subkey in zip(guess_value, subkeys)]
+    else:
+        return Uniform(minval, np.inf, guess_value, key)
```

Comment:
Same as above, here. (I'm a little confused as to what this is doing, since it seems to assume keys have a colon in them, when my impression of the new version was that keys had no colon.

It might be a little clearer if you split out the code that gets the guess value?? Or maybe I'm missing something completely.

Reply:
Colons are still used by composite scatterers to denote their individual components.

Automated detection of fitting parameters from scatterers got complicated when scatterers forgot how to interpret themselves. Specifically, the combination of Spheres objects and trying to fit one or more 'center' coordinates ended up producing a bunch of cases that need to be handled. This is why I had suggested not doing the automatic detection for composite scatterers. 

Note that all of this parameter handling in interface.py is just to come up with a Scatterer object containing prior objects for the case where a user calls hp.fit(data, scatterer). More advanced users will define their own Scatterer with prior objects and a Model.